File: app/src/main/java/com/igordanilchik/daggertest/ui/fragment/OfferFragment.java
package com.igordanilchik.daggertest.ui.fragment;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.igordanilchik.daggertest.DaggerApplication;
import com.igordanilchik.daggertest.R;
import com.igordanilchik.daggertest.data.Repository;
import com.igordanilchik.daggertest.model.Catalogue;
import com.igordanilchik.daggertest.model.Offer;
import com.igordanilchik.daggertest.ui.activity.MainActivity;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import rx.Observable;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import rx.subscriptions.Subscriptions;

public class OfferFragment extends Fragment {
    private static final String LOG_TAG = OfferFragment.class.getSimpleName();

    @BindView(R.id.card_image)
    ImageView image;
    @BindView(R.id.card_title)
    TextView title;
    @BindView(R.id.card_price)
    TextView price;
    @BindView(R.id.card_weight)
    TextView weight;
    @BindView(R.id.card_description)
    TextView description;

    private Unbinder unbinder;

    @Inject
    Repository repository;

    @NonNull
    Subscription subscription = Subscriptions.empty();

    private int offerId;

    @NonNull
    public static OfferFragment newInstance(int offerId) {
        Bundle args = new Bundle();
        args.putInt(MainActivity.ARG_OFFER_ID, offerId);

        OfferFragment f = new OfferFragment();
        f.setArguments(args);
        return f;
    }

    @Override
    public void onCreate(final @Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        DaggerApplication.get(this.getContext())
                .getRepositoryComponent()
                .inject(this);

        Bundle bundle = getArguments();
        if (bundle != null && bundle.containsKey(MainActivity.ARG_OFFER_ID)) {
            offerId = bundle.getInt(MainActivity.ARG_OFFER_ID);
        }
    }

    @Override
    public View onCreateView(@NonNull final LayoutInflater inflater, final @Nullable ViewGroup container, final @Nullable Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.fragment_offer, container, false);
        this.unbinder = ButterKnife.bind(this, view);

        requestData();
        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();

        subscription.unsubscribe();
        Log.d(LOG_TAG, "Observer unsubscribed");
        unbinder.unbind();
    }


    private void requestData() {
        subscription.unsubscribe();
        Log.d(LOG_TAG, "Observer unsubscribed");

        subscription = repository.getCatalogue(false)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(Catalogue::getShop)
                .filter(shop -> shop != null && shop.getOffers() != null)
                .flatMap(shop -> Observable.just(shop.getOffers()))
                .flatMap(Observable::from)
                .filter(offer -> offer.getId() == offerId)
                .subscribe(this::updateContent);

        Log.d(LOG_TAG, "Observer subscribed");
    }

    private void updateContent(@Nullable Offer offer) {
        if (offer != null) {
            title.setText(offer.getName());
            price.setText(getString(R.string.offer_price, offer.getPrice()));

            if (offer.getParam() != null) {
                String weightStr = offer.getParam().get(this.getString(R.string.param_name_weight));
                if (weightStr != null) {
                    weight.setText(getString(R.string.offer_weight, weightStr));
                }
            }

            String url = offer.getPictureUrl();
            if (url != null && !url.isEmpty()) {
                Glide.with(this)
                        .load(url)
                        .fitCenter()
                        .placeholder(ContextCompat.getDrawable(getContext(), R.drawable.ic_image_black_24dp))
                        .crossFade()
                        .diskCacheStrategy(DiskCacheStrategy.ALL)
                        .into(image);
            } else {
                image.setVisibility(View.GONE);
            }

            String descriptionText = offer.getDescription();
            description.setText(descriptionText);
        }
    }
}

File: app/src/main/java/com/igordanilchik/daggertest/flows/offer/view/OfferFragment.kt
package com.igordanilchik.daggertest.flows.offer.view

import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v4.content.ContextCompat
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import butterknife.BindView
import com.arellomobile.mvp.presenter.InjectPresenter
import com.arellomobile.mvp.presenter.ProvidePresenter
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.igordanilchik.daggertest.R
import com.igordanilchik.daggertest.common.mvp.view.BaseFragment
import com.igordanilchik.daggertest.data.Offers
import com.igordanilchik.daggertest.flows.offer.builder.OfferModule
import com.igordanilchik.daggertest.flows.offer.model.OfferSupplier
import com.igordanilchik.daggertest.flows.offer.presenter.OfferPresenter
import com.igordanilchik.daggertest.ui.activity.MainActivity

/**
 * @author Igor Danilchik
 */
class OfferFragment : BaseFragment(), OfferView {

    @BindView(R.id.card_image)
    lateinit var image: ImageView
    @BindView(R.id.card_title)
    lateinit var title: TextView
    @BindView(R.id.card_price)
    lateinit var price: TextView
    @BindView(R.id.card_weight)
    lateinit var weight: TextView
    @BindView(R.id.card_description)
    lateinit var description: TextView
    @BindView(R.id.linear_layout)
    lateinit var linearLayout: LinearLayout

    @InjectPresenter
    lateinit var presenter: OfferPresenter

    override val layoutResID = R.layout.fragment_offer

    override fun showOffer(offer: Offers.Offer) {
        title.text = offer.name
        price.text = getString(R.string.offer_price, offer.price)

        getParamByKey(offer, getString(R.string.param_name_weight)).let {
            weight.text = getString(R.string.offer_weight, it)
        }

        offer.picture?.let { url ->
            url.takeIf { it.isNotEmpty() }?.let {
                Glide.with(this)
                        .load(it)
                        .fitCenter()
                        .placeholder(ContextCompat.getDrawable(context, R.drawable.ic_image_black_24dp))
                        .crossFade()
                        .diskCacheStrategy(DiskCacheStrategy.ALL)
                        .into(image)
            }
        } ?: run { image.visibility = View.GONE }

        description.text = offer.description
    }

    private fun getParamByKey(list: Offers.Offer?, key: String): String? = list?.param?.find { it.name == key }?.value

    override fun showProgress() {

    }

    override fun hideProgress() {

    }

    override fun showError(throwable: Throwable) {
        Snackbar.make(linearLayout, "Error: " + throwable.message, Snackbar.LENGTH_LONG)
                .show()
    }

    @ProvidePresenter
    fun providePresenter(): OfferPresenter {
        val bundle = arguments ?: Bundle()
        val supplier = OfferSupplier(id = bundle.getInt(MainActivity.ARG_OFFER_ID))

        return appComponent().plusOfferComponent(OfferModule(supplier)).presenter()
    }


}